src/pytorch_lightning/trainer/trainer.py

```diff
@@ -637,6 +641,13 @@ def _call_and_handle_interrupt(self, trainer_fn: Callable, *args: Any, **kwargs:
                 return self.strategy.launcher.launch(trainer_fn, *args, trainer=self, **kwargs)
             else:
                 return trainer_fn(*args, **kwargs)
+
+        except _TunerExitException as exception:
+            self.state.status = TrainerStatus.FINISHED
+            self._call_callback_hooks("on_exception", exception)
+            self._teardown()
+            self.state.stage = None
```

Comment:
```suggestion
            self._call_teardown_hook()
            self._teardown()
            self.state.stage = None
            self.state.status = TrainerStatus.FINISHED
```
need to think more. For now can we do this? Even if we do `loop.shoud_stop`, it will update this state of the loop and during the next run of let's say `trainer.fit/validate`, loop might not start because the last exit will have `loop.shoud_stop=True`.

Reply:
Yes. I think the tuner callback should completely reset the loop states on teardown.

Reply:
rather I'd say it should restore the loop state which it does while restoring the params.

but if we do:
> Could we avoid using an exception and instead do loop.should_stop?

then this loop state `loop.should_stop` will be True and during the next run, it will exit without running anything
```
trainer.tune() -> sets `loop.should_stop = False`
trainer.fit() -> might not loop

```

Reply:
Just as the Callback could signal to stop, it could also unsignal it on teardown, no?